src/test/java/org/jitsi/meet/test/MalleusJitsificus.java

```diff
@@ -170,69 +176,61 @@ public void testMain(
         String[] regions, float blipMaxDisruptedPct)
         throws Exception
     {
-        ParticipantThread[] runThreads = new ParticipantThread[numberOfParticipants];
+        Future<?>[] futures = new Future[numberOfParticipants];
 
         bridgeSelectionCountDownLatch = new CountDownLatch(numberOfParticipants);
 
+        ExecutorService pool = Executors.newFixedThreadPool(numberOfParticipants);
+
         boolean disruptBridges = blipMaxDisruptedPct > 0;
         for (int i = 0; i < numberOfParticipants; i++)
         {
-         runThreads[i]
-                = new ParticipantThread(
+            futures[i] = pool.submit(new MalleusTask(
                 i,
                 url.copy(),
                 durationMs,
                 i * joinDelayMs,
                 i >= numSenders /* no video */,
                 i >= numAudioSenders /* no audio */,
                 regions == null ? null : regions[i % regions.length],
-                disruptBridges ? 0 : DISABLE_FAILURE_DETECTION
-             );
-
-            runThreads[i].start();
+                disruptBridges
+            ));
         }
 
+        List<Throwable> errors = new ArrayList<>();
         if (disruptBridges)
         {
             try
             {
-                disruptBridges(blipMaxDisruptedPct, durationMs / 1000, runThreads);
+                disruptBridges(blipMaxDisruptedPct, durationMs / 1000);
             }
             catch (Exception e)
             {
-                throw new Exception("Failed to disrupt the bridges.");
-            }
-            finally
-            {
-                for (ParticipantThread t : runThreads)
-                {
-                    if (t != null)
-                    {
-                        t.join();
-                    }
-                }
+                // TODO we should cancel the futures since test has failed
+                errors.add(new Exception("Failed to disrupt the bridges.", e));
             }
         }
 
-        int minFailureTolerance = Integer.MAX_VALUE;
```

Comment:
The logic that checks if the participant failed "too much" was moved into the participant thread code and changed into throwing an exception (that is now handled).